sys/log/full/src/log_fcb2.c

```diff
@@ -0,0 +1,866 @@
+        return len;
+    } else {
+        return 0;
+    }
+}
+
+static int
+log_fcb2_read_mbuf(struct log *log, void *dptr, struct os_mbuf *om,
+                   uint16_t off, uint16_t len)
+{
+    struct fcb_entry *loc;
+    uint8_t data[128];
+    uint16_t read_len;
+    uint16_t rem_len;
+    int rc;
+
+    loc = (struct fcb_entry *)dptr;
+
+    if (off + len > loc->fe_data_len) {
+        len = loc->fe_data_len - off;
+    }
+
+    rem_len = len;
+
+    while (rem_len > 0) {
+        read_len = min(rem_len, sizeof(data));
+        rc = fcb_read(loc, off, data, read_len);
+        if (rc) {
+            goto done;
+        }
+
+        rc = os_mbuf_append(om, data, read_len);
+        if (rc) {
+            goto done;
+        }
+
+        rem_len -= read_len;
+        off += read_len;
+    }
+
+done:
+    return len - rem_len;
+}
+
+static int
+log_fcb2_walk(struct log *log, log_walk_func_t walk_func,
+              struct log_offset *log_off)
+{
+    struct fcb *fcb;
+    struct fcb_log *fcb_log;
+    struct fcb_entry loc;
+    int rc;
+
+    fcb_log = log->l_arg;
+    fcb = &fcb_log->fl_fcb;
+
+    /* Locate the starting point of the walk. */
+    rc = log_fcb2_find_gte(log, log_off, &loc);
+    switch (rc) {
+    case 0:
+        /* Found a starting point. */
+        break;
+    case SYS_ENOENT:
+        /* No entries match the offset criteria; nothing to walk. */
+        return 0;
+    default:
+        return rc;
+    }
+
+#if MYNEWT_VAL(LOG_FCB_BOOKMARKS)
+    /* If a minimum index was specified (i.e., we are not just retrieving the
+     * last entry), add a bookmark pointing to this walk's start location.
+     */
+    if (log_off->lo_ts >= 0) {
+        fcb_log_add_bmark(fcb_log, &loc, log_off->lo_index);
+    }
+#endif
+
+    do {
+        rc = walk_func(log, log_off, &loc, loc.fe_data_len);
+        if (rc != 0) {
+            return rc;
+        }
+    } while (fcb_getnext(fcb, &loc) == 0);
+
+    return 0;
+}
+
+static int
+log_fcb2_flush(struct log *log)
+{
+    struct fcb_log *fcb_log;
+    struct fcb *fcb;
+
+    fcb_log = (struct fcb_log *)log->l_arg;
+    fcb = &fcb_log->fl_fcb;
+
+#if MYNEWT_VAL(LOG_FCB_BOOKMARKS)
+    fcb_log_clear_bmarks(fcb_log);
+#endif
+
+    return fcb_clear(fcb);
+}
+
+static int
+log_fcb2_registered(struct log *log)
+{
+    struct fcb *fcb;
+    int i;
+    struct fcb_log *fl;
+#if MYNEWT_VAL(LOG_STORAGE_WATERMARK)
+#if MYNEWT_VAL(LOG_PERSIST_WATERMARK)
+    struct fcb_entry loc;
+#endif
+#endif
+
+    fl = (struct fcb_log *)log->l_arg;
+    fcb = &fl->fl_fcb;
+
+    for (i = 0; i < fcb->f_range_cnt; i++) {
+        if (fcb->f_ranges[i].fsr_align > LOG_FCB2_MAX_ALIGN) {
+            return SYS_ENOTSUP;
+        }
+    }
+
+#if MYNEWT_VAL(LOG_STORAGE_WATERMARK)
+#if MYNEWT_VAL(LOG_PERSIST_WATERMARK)
+    /* Set watermark to first element */
+    memset(&loc, 0, sizeof(loc));
+
+    if (fcb_getnext(fcb, &loc)) {
+        fl->fl_watermark_off = loc.fe_area->fa_off + loc.fe_elem_off;
```

Comment:
filed **fe_area** and **fe_elem_off** not present in **FCB2** build

Reply:
Watermarking was not yet supported.